lib/fortnightly_performance_report.rb

```diff
@@ -0,0 +1,35 @@
+class FortnightlyPerformanceReport
+  def initialize(model_class, prison_name, year)
+    @model = model_class
+    @prison_name = prison_name
+    @year = year
+  end
+
+  def performance
+    @model.connection.execute(%Q{
+WITH percentiles AS (
+  SELECT requested_at,
+  EXTRACT(week FROM requested_at)::integer / 2 AS fortnight,
+         end_to_end_time,
+         cume_dist() OVER (PARTITION BY EXTRACT(week FROM requested_at)::integer / 2
+                           ORDER BY end_to_end_time)
+  FROM visit_metrics_entries
+  WHERE end_to_end_time IS NOT NULL
+  AND prison_name = '#{@prison_name}' AND EXTRACT(isoyear FROM requested_at) = #{@year}
+  ORDER BY fortnight)
+SELECT MIN(DATE_TRUNC('week', requested_at))::date AS x, MIN(end_to_end_time) AS y
+FROM percentiles
+WHERE cume_dist >= 0.95
+GROUP BY fortnight
+ORDER BY x})
+  end
+
+  def volume
+    @model.connection.execute(%Q{
+SELECT MIN(DATE_TRUNC('week', requested_at))::date AS x, COUNT(*) AS y FROM visit_metrics_entries
+WHERE prison_name = '#{@prison_name}' AND EXTRACT(isoyear FROM requested_at) = #{@year}
```

Comment:
Can we replace the string interpolations with parameters? I know this isn't connected to anything user-facing directly, but it might still leave us vulnerable to SQL injection… Something like:

```ruby
query = connection.raw_connection.prepare(%Q{
SELECT …
WHERE prison_name=? AND EXTRACT(isoyear FROM requested_at) = ?
…
})
query.execute(@prison_name, @year)
```

That way, ActiveRecord sanitises and cleans all the input for us so the variables can't be used to escape the statement and trash our DB.